File: go/knowledge/ingest.go
// SPDX-License-Identifier: Apache-2.0

package knowledge

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
	"unicode"
	"unicode/utf8"

	pdfreader "github.com/ledongthuc/pdf"

	"github.com/jeffs-brain/memory/go/brain"
)

// Content-type constants for the supported raw formats. Used by the
// ingest pipeline to route a [IngestRequest] to the correct extractor.
const (
	contentTypeMarkdown = "text/markdown"
	contentTypeText     = "text/plain"
	contentTypeHTML     = "text/html"
	contentTypePDF      = "application/pdf"
	contentTypeJSON     = "application/json"
	contentTypeYAML     = "application/x-yaml"
)

// maxReadBytes caps the body we accept from any single source. Matches
// jeff's text/binary ingest limits rolled into one conservative ceiling
// so an oversized URL or local file cannot exhaust memory.
const maxReadBytes = 50 * 1024 * 1024

// defaultHTTPTimeout is the per-request wall clock budget for
// [defaultFetcher]. Matches jeff's ingest.go default.
const defaultHTTPTimeout = 30 * time.Second

// htmlTagPattern strips every HTML tag from a document so the plain
// extractor never emits markup to the index. Matches both open and
// self-closing tags.
var htmlTagPattern = regexp.MustCompile(`(?s)<[^>]+>`)

// scriptPattern and stylePattern drop <script> and <style> blocks
// wholesale before the generic tag stripper runs so the index never
// captures JavaScript or stylesheet text as content. Go's RE2 does not
// support backreferences, so each block gets its own explicit pattern.
var (
	scriptPattern = regexp.MustCompile(`(?is)<script[^>]*>.*?</\s*script\s*>`)
	stylePattern  = regexp.MustCompile(`(?is)<style[^>]*>.*?</\s*style\s*>`)
)

// defaultFetcher is the production HTTP fetcher used by [Base.IngestURL]
// when the caller does not supply an override.
type defaultFetcher struct{}

// Fetch implements [Fetcher].
func (defaultFetcher) Fetch(ctx context.Context, rawURL string) ([]byte, string, error) {
	client := &http.Client{Timeout: defaultHTTPTimeout}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, rawURL, nil)
	if err != nil {
		return nil, "", fmt.Errorf("knowledge: building request: %w", err)
	}
	req.Header.Set("Accept", "text/plain, text/markdown, text/html, application/pdf")
	resp, err := client.Do(req)
	if err != nil {
		return nil, "", fmt.Errorf("knowledge: fetching %s: %w", rawURL, err)
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, "", fmt.Errorf("knowledge: fetch %s: HTTP %d", rawURL, resp.StatusCode)
	}
	body, err := io.ReadAll(io.LimitReader(resp.Body, maxReadBytes))
	if err != nil {
		return nil, "", fmt.Errorf("knowledge: reading response: %w", err)
	}
	ctype := resp.Header.Get("Content-Type")
	return body, ctype, nil
}

// Ingest implements [Base].
func (k *kbase) Ingest(ctx context.Context, req IngestRequest) (IngestResponse, error) {
	if err := k.requireStore(); err != nil {
		return IngestResponse{}, err
	}
	start := time.Now()

	raw, ctype, sourceLabel, err := k.readRequestBody(req)
	if err != nil {
		return IngestResponse{}, err
	}

	extracted, err := extractPlain(raw, ctype, filepath.Ext(strings.ToLower(req.Path)))
	if err != nil {
		return IngestResponse{}, err
	}

	doc := buildDocument(req, ctype, sourceLabel, extracted, raw)
	if err := k.writeDocument(ctx, doc); err != nil {
		return IngestResponse{}, err
	}

	// Compile inline so the index reflects the ingest immediately.
	chunks, err := k.chunkAndIndex(ctx, doc)
	if err != nil {
		return IngestResponse{}, err
	}

	return IngestResponse{
		DocumentID: doc.ID,
		Path:       doc.Path,
		ChunkCount: chunks,
		Bytes:      doc.Bytes,
		TookMs:     time.Since(start).Milliseconds(),
	}, nil
}

// IngestURL implements [Base]. Fetches the URL through the configured
// [Fetcher], routes the body through the plain-text extractor, and
// persists the result.
func (k *kbase) IngestURL(ctx context.Context, rawURL string) (IngestResponse, error) {
	if err := k.requireStore(); err != nil {
		return IngestResponse{}, err
	}
	parsed, err := normaliseURL(rawURL)
	if err != nil {
		return IngestResponse{}, err
	}
	_, _, fetcher := k.snapshot()
	body, ctype, err := fetcher.Fetch(ctx, parsed)
	if err != nil {
		return IngestResponse{}, err
	}
	return k.Ingest(ctx, IngestRequest{
		BrainID:     k.brainID,
		Path:        parsed,
		ContentType: ctype,
		Content:     bytes.NewReader(body),
	})
}

// readRequestBody loads the request body. When req.Content is set it
// takes precedence; otherwise req.Path is opened from the local
// filesystem. Returns the raw bytes, the detected content type, and a
// source label suitable for frontmatter.
func (k *kbase) readRequestBody(req IngestRequest) ([]byte, string, string, error) {
	ctype := strings.ToLower(strings.TrimSpace(req.ContentType))
	// Normalise the charset suffix (e.g. "text/html; charset=utf-8")
	// so downstream switches work on the base media type alone.
	if idx := strings.Index(ctype, ";"); idx >= 0 {
		ctype = strings.TrimSpace(ctype[:idx])
	}

	if req.Content != nil {
		data, err := io.ReadAll(io.LimitReader(req.Content, maxReadBytes))
		if err != nil {
			return nil, "", "", fmt.Errorf("knowledge: reading content: %w", err)
		}
		if ctype == "" {
			ctype = detectContentType(req.Path, data)
		}
		return data, ctype, firstNonEmpty(req.Path, "inline"), nil
	}

	if strings.TrimSpace(req.Path) == "" {
		return nil, "", "", fmt.Errorf("knowledge: either Content or Path required")
	}

	if _, parseErr := url.ParseRequestURI(req.Path); parseErr == nil && strings.Contains(req.Path, "://") {
		return nil, "", "", fmt.Errorf("knowledge: use IngestURL for %s", req.Path)
	}

	abs, err := filepath.Abs(req.Path)
	if err != nil {
		return nil, "", "", fmt.Errorf("knowledge: resolving path: %w", err)
	}
	info, err := os.Stat(abs)
	if err != nil {
		return nil, "", "", fmt.Errorf("knowledge: stat %s: %w", abs, err)
	}
	if info.IsDir() {
		return nil, "", "", fmt.Errorf("knowledge: %s is a directory", abs)
	}
	if info.Size() > maxReadBytes {
		return nil, "", "", fmt.Errorf("knowledge: %s exceeds %d byte limit", abs, maxReadBytes)
	}
	data, err := os.ReadFile(abs)
	if err != nil {
		return nil, "", "", fmt.Errorf("knowledge: reading %s: %w", abs, err)
	}
	if ctype == "" {
		ctype = detectContentType(abs, data)
	}
	return data, ctype, abs, nil
}

// detectContentType picks a content type from the file extension first,
// falling back to http.DetectContentType for binary blobs without a
// useful suffix.
func detectContentType(p string, data []byte) string {
	ext := strings.ToLower(filepath.Ext(p))
	switch ext {
	case ".md", ".markdown":
		return contentTypeMarkdown
	case ".txt", ".text", ".log":
		return contentTypeText
	case ".html", ".htm":
		return contentTypeHTML
	case ".pdf":
		return contentTypePDF
	case ".json":
		return contentTypeJSON
	case ".yaml", ".yml":
		return contentTypeYAML
	}
	if len(data) > 4 && string(data[:4]) == "%PDF" {
		return contentTypePDF
	}
	sniff := http.DetectContentType(data)
	if strings.HasPrefix(sniff, "text/html") {
		return contentTypeHTML
	}
	if strings.HasPrefix(sniff, "text/") {
		return contentTypeText
	}
	return sniff
}

// extractPlain routes raw bytes to the correct extractor. Markdown and
// plain text pass through untouched; HTML is stripped; PDF is decoded
// via ledongthuc/pdf. Everything else is rejected.
func extractPlain(raw []byte, ctype, ext string) (string, error) {
	base := strings.ToLower(strings.TrimSpace(ctype))
	if idx := strings.Index(base, ";"); idx >= 0 {
		base = strings.TrimSpace(base[:idx])
	}
	switch base {
	case contentTypeMarkdown, contentTypeText, contentTypeJSON, contentTypeYAML:
		if !utf8.Valid(raw) {
			return "", fmt.Errorf("knowledge: content is not valid UTF-8")
		}
		return string(raw), nil
	case contentTypeHTML:
		return stripHTML(raw), nil
	case contentTypePDF:
		return extractPDF(raw)
	}
	if ext == ".pdf" || (len(raw) > 4 && string(raw[:4]) == "%PDF") {
		return extractPDF(raw)
	}
	if strings.HasPrefix(base, "text/") {
		if !utf8.Valid(raw) {
			return "", fmt.Errorf("knowledge: content is not valid UTF-8")
		}
		return string(raw), nil
	}
	return "", fmt.Errorf("knowledge: unsupported content-type %q", ctype)
}

// stripHTML drops scripts, styles, and tags to produce a plain-text
// extract suitable for indexing.
func stripHTML(raw []byte) string {
	s := string(raw)
	s = scriptPattern.ReplaceAllString(s, " ")
	s = stylePattern.ReplaceAllString(s, " ")
	s = htmlTagPattern.ReplaceAllString(s, " ")
	s = strings.ReplaceAll(s, "&nbsp;", " ")
	s = strings.ReplaceAll(s, "&amp;", "&")
	s = strings.ReplaceAll(s, "&lt;", "<")
	s = strings.ReplaceAll(s, "&gt;", ">")
	s = strings.ReplaceAll(s, "&quot;", `"`)
	return collapseWhitespace(s)
}

// extractPDF runs the ledongthuc/pdf GetPlainText path across every
// page and returns the concatenated text.
func extractPDF(raw []byte) (string, error) {
	if len(raw) == 0 {
		return "", fmt.Errorf("knowledge: empty pdf body")
	}
	reader := bytes.NewReader(raw)
	r, err := pdfreader.NewReader(reader, int64(len(raw)))
	if err != nil {
		return "", fmt.Errorf("knowledge: opening pdf: %w", err)
	}
	var b strings.Builder
	pages := r.NumPage()
	for i := 1; i <= pages; i++ {
		p := r.Page(i)
		if p.V.IsNull() {
			continue
		}
		text, err := p.GetPlainText(nil)
		if err != nil {
			// Non-fatal; return what we have so far. PDF extractors
			// commonly trip over malformed embedded fonts mid-document.
			continue
		}
		if text != "" {
			if b.Len() > 0 {
				b.WriteString("\n\n")
			}
			b.WriteString(text)
		}
	}
	return collapseWhitespace(b.String()), nil
}

// buildDocument assembles a [Document] from a request plus the already-
// extracted plain text. Title, summary, and tags prefer frontmatter
// when present, otherwise fall back to the request hints.
func buildDocument(req IngestRequest, ctype, sourceLabel, extracted string, raw []byte) *Document {
	now := time.Now().UTC()
	title := strings.TrimSpace(req.Title)
	var summary string
	tags := append([]string{}, req.Tags...)

	fmParsed, body := ParseFrontmatter(extracted)
	if title == "" {
		switch {
		case fmParsed.Title != "":
			title = fmParsed.Title
		case fmParsed.Name != "":
			title = fmParsed.Name
		default:
			title = deriveTitle(body, sourceLabel)
		}
	}
	summary = fmParsed.Summary
	if summary == "" {
		summary = fmParsed.Description
	}
	if len(fmParsed.Tags) > 0 {
		tags = append(tags, fmParsed.Tags...)
	}

	slug := slugify(title)
	if slug == "" {
		slug = hashSlug(raw)
	}

	docID := hashSlug(append([]byte(slug+":"), raw...))
	logical := brain.RawDocument(slug)

	return &Document{
		ID:          docID,
		Path:        logical,
		Title:       title,
		Source:      sourceLabel,
		ContentType: ctype,
		Tags:        dedupeStrings(tags),
		Summary:     summary,
		Body:        strings.TrimSpace(body),
		Bytes:       len(raw),
		Ingested:    now,
		Modified:    now,
	}
}

// writeDocument persists the document under raw/documents/<slug>.md as a
// canonical markdown file with a regenerated YAML frontmatter header.
func (k *kbase) writeDocument(ctx context.Context, doc *Document) error {
	fm := buildFrontmatterYAML(doc)
	payload := fm + "\n\n" + doc.Body + "\n"
	return k.store.Write(ctx, doc.Path, []byte(payload))
}

// buildFrontmatterYAML emits the canonical frontmatter block written
// alongside every ingested document. British English throughout, no em
// dashes, consistent quoting.
func buildFrontmatterYAML(doc *Document) string {
	var b strings.Builder
	b.WriteString("---\n")
	writeYAMLString(&b, "title", doc.Title)
	if doc.Summary != "" {
		writeYAMLString(&b, "summary", doc.Summary)
	}
	writeYAMLString(&b, "source", doc.Source)
	writeYAMLString(&b, "source_type", routeSourceType(doc.ContentType))
	writeYAMLString(&b, "ingested", doc.Ingested.Format(time.RFC3339))
	writeYAMLString(&b, "modified", doc.Modified.Format(time.RFC3339))
	if len(doc.Tags) > 0 {
		b.WriteString("tags:\n")
		for _, t := range doc.Tags {
			b.WriteString("  - ")
			b.WriteString(t)
			b.WriteString("\n")
		}
	}
	b.WriteString("---")
	return b.String()
}

// writeYAMLString emits a single quoted key/value pair.
func writeYAMLString(b *strings.Builder, key, value string) {
	b.WriteString(key)
	b.WriteString(`: `)
	b.WriteString(quoteYAML(value))
	b.WriteString("\n")
}

// quoteYAML wraps value in double quotes and escapes embedded quotes.
func quoteYAML(value string) string {
	value = strings.ReplaceAll(value, `\`, `\\`)
	value = strings.ReplaceAll(value, `"`, `\"`)
	return `"` + value + `"`
}

// routeSourceType maps a content type to the frontmatter source_type
// tag. Used to keep downstream filters consistent across ingest paths.
func routeSourceType(ctype string) string {
	switch ctype {
	case contentTypeMarkdown:
		return "markdown"
	case contentTypeText:
		return "text"
	case contentTypeHTML:
		return "html"
	case contentTypePDF:
		return "pdf"
	case contentTypeJSON:
		return "json"
	case contentTypeYAML:
		return "yaml"
	}
	return "document"
}

// deriveTitle picks a best-effort title from the body when the caller
// provides none. Prefers the first `# heading`, then the first
// non-empty line, then the file basename.
func deriveTitle(body, fallback string) string {
	for _, line := range strings.Split(body, "\n") {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "# ") {
			return strings.TrimSpace(strings.TrimPrefix(line, "# "))
		}
	}
	for _, line := range strings.Split(body, "\n") {
		line = strings.TrimSpace(line)
		if line != "" {
			if len(line) > 120 {
				line = line[:120]
			}
			return line
		}
	}
	base := filepath.Base(fallback)
	base = strings.TrimSuffix(base, filepath.Ext(base))
	if base == "" || base == "." {
		return "untitled"
	}
	return base
}

// hashSlug is a deterministic fallback slug built from a SHA-256 sum
// truncated to 12 hex characters. Used when slugify collapses to empty.
func hashSlug(data []byte) string {
	sum := sha256.Sum256(data)
	return hex.EncodeToString(sum[:])[:12]
}

// slugify mirrors jeff's helper: lowercase, hyphen-separated, capped at
// 60 characters. Unicode letters and digits survive; everything else
// becomes a hyphen.
func slugify(title string) string {
	s := strings.ToLower(strings.TrimSpace(title))
	var b strings.Builder
	for _, r := range s {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			b.WriteRune(r)
		} else {
			b.WriteRune('-')
		}
	}
	slug := collapseHyphens(b.String())
	slug = strings.Trim(slug, "-")
	if len(slug) > 60 {
		slug = slug[:60]
		slug = strings.TrimRight(slug, "-")
	}
	return slug
}

// collapseHyphens reduces runs of hyphens to a single hyphen.
func collapseHyphens(s string) string {
	var b strings.Builder
	var prevDash bool
	for _, r := range s {
		if r == '-' {
			if prevDash {
				continue
			}
			prevDash = true
			b.WriteRune(r)
			continue
		}
		prevDash = false
		b.WriteRune(r)
	}
	return b.String()
}

// collapseWhitespace trims and collapses runs of whitespace to a single
// space, preserving paragraph breaks where possible.
func collapseWhitespace(s string) string {
	paragraphs := strings.Split(s, "\n\n")
	for i, p := range paragraphs {
		paragraphs[i] = strings.Join(strings.Fields(p), " ")
	}
	joined := strings.Join(paragraphs, "\n\n")
	return strings.TrimSpace(joined)
}

// normaliseURL ensures a URL has a scheme and is well-formed.
func normaliseURL(raw string) (string, error) {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return "", fmt.Errorf("knowledge: empty URL")
	}
	if !strings.Contains(raw, "://") {
		raw = "https://" + raw
	}
	parsed, err := url.Parse(raw)
	if err != nil {
		return "", fmt.Errorf("knowledge: parsing URL: %w", err)
	}
	if parsed.Host == "" {
		return "", fmt.Errorf("knowledge: URL missing host")
	}
	return parsed.String(), nil
}

// firstNonEmpty returns the first non-empty argument.
func firstNonEmpty(ss ...string) string {
	for _, s := range ss {
		if strings.TrimSpace(s) != "" {
			return s
		}
	}
	return ""
}

// dedupeStrings removes duplicate entries while preserving order.
func dedupeStrings(in []string) []string {
	seen := make(map[string]struct{}, len(in))
	out := make([]string, 0, len(in))
	for _, s := range in {
		key := strings.TrimSpace(s)
		if key == "" {
			continue
		}
		if _, ok := seen[key]; ok {
			continue
		}
		seen[key] = struct{}{}
		out = append(out, key)
	}
	return out
}

File: sdks/py/src/jeffs_brain_memory/knowledge/ingest.py
# SPDX-License-Identifier: Apache-2.0
"""Ingest routing — markdown, plain text, HTML, PDF, and URLs.

Ported from ``go/knowledge/ingest.go``. Every ingested document
lands under ``raw/documents/<slug>.md`` with a regenerated YAML
frontmatter header so downstream search indexes see a canonical shape.
"""

from __future__ import annotations

import hashlib
import io
import re
import unicodedata
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import BinaryIO, Protocol, runtime_checkable
from urllib.parse import urlparse

import httpx

from ..path import BrainPath, DocumentID
from .frontmatter import Frontmatter, parse_frontmatter
from .types import Document, IngestRequest

__all__ = [
    "CONTENT_TYPE_MARKDOWN",
    "CONTENT_TYPE_TEXT",
    "CONTENT_TYPE_HTML",
    "CONTENT_TYPE_PDF",
    "CONTENT_TYPE_JSON",
    "CONTENT_TYPE_YAML",
    "MAX_READ_BYTES",
    "DEFAULT_HTTP_TIMEOUT",
    "RAW_DOCUMENTS_PREFIX",
    "Fetcher",
    "DefaultFetcher",
    "build_document",
    "build_frontmatter_yaml",
    "detect_content_type",
    "extract_plain",
    "normalise_url",
    "raw_document_path",
    "slugify",
    "strip_html",
]


# Content-type constants for the raw formats we accept. Matches the Go
# reference so cross-SDK callers can compare against the same strings.
CONTENT_TYPE_MARKDOWN = "text/markdown"
CONTENT_TYPE_TEXT = "text/plain"
CONTENT_TYPE_HTML = "text/html"
CONTENT_TYPE_PDF = "application/pdf"
CONTENT_TYPE_JSON = "application/json"
CONTENT_TYPE_YAML = "application/x-yaml"

# Cap for any single source body. Matches jeff's conservative ceiling so
# an oversized URL or file cannot exhaust memory.
MAX_READ_BYTES = 50 * 1024 * 1024

# Per-request wall clock for :class:`DefaultFetcher`. Matches Go.
DEFAULT_HTTP_TIMEOUT = 30.0

# Logical prefix for every ingest. Mirrors the Go ``RawDocumentsPrefix``
# helper so search indexes that walk ``raw/documents/`` pick up
# everything this module writes.
RAW_DOCUMENTS_PREFIX = "raw/documents"

# Precompiled regex patterns for the HTML stripper. The script/style
# patterns run first so neither JavaScript nor stylesheet text ends up in
# the extracted body. Python supports backreferences but we keep the
# trio of patterns here to match the Go RE2 implementation line for line.
_SCRIPT_RE = re.compile(r"<script[^>]*>.*?</\s*script\s*>", re.IGNORECASE | re.DOTALL)
_STYLE_RE = re.compile(r"<style[^>]*>.*?</\s*style\s*>", re.IGNORECASE | re.DOTALL)
_TAG_RE = re.compile(r"<[^>]+>", re.DOTALL)


@runtime_checkable
class Fetcher(Protocol):
    """Abstract HTTP fetcher so :func:`ingest_url` stays testable."""

    async def fetch(self, url: str) -> tuple[bytes, str]:
        """Return the response body and content-type for ``url``."""
        ...


@dataclass(slots=True)
class DefaultFetcher:
    """Production HTTP fetcher backed by :mod:`httpx`."""

    timeout: float = DEFAULT_HTTP_TIMEOUT

    async def fetch(self, url: str) -> tuple[bytes, str]:
        headers = {
            "Accept": "text/plain, text/markdown, text/html, application/pdf",
        }
        async with httpx.AsyncClient(timeout=self.timeout, follow_redirects=True) as client:
            resp = await client.get(url, headers=headers)
        if resp.status_code < 200 or resp.status_code >= 300:
            raise ValueError(f"knowledge: fetch {url}: HTTP {resp.status_code}")
        body = resp.content[:MAX_READ_BYTES]
        ctype = resp.headers.get("content-type", "")
        return body, ctype


def raw_document_path(slug: str) -> BrainPath:
    """Return the ``raw/documents/<slug>.md`` logical path."""
    if not slug:
        slug = "untitled"
    return BrainPath(f"{RAW_DOCUMENTS_PREFIX}/{slug}.md")


def detect_content_type(path: str, data: bytes) -> str:
    """Pick a content-type from extension first, magic bytes second."""
    ext = _extension(path).lower()
    ext_map = {
        ".md": CONTENT_TYPE_MARKDOWN,
        ".markdown": CONTENT_TYPE_MARKDOWN,
        ".txt": CONTENT_TYPE_TEXT,
        ".text": CONTENT_TYPE_TEXT,
        ".log": CONTENT_TYPE_TEXT,
        ".html": CONTENT_TYPE_HTML,
        ".htm": CONTENT_TYPE_HTML,
        ".pdf": CONTENT_TYPE_PDF,
        ".json": CONTENT_TYPE_JSON,
        ".yaml": CONTENT_TYPE_YAML,
        ".yml": CONTENT_TYPE_YAML,
    }
    if ext in ext_map:
        return ext_map[ext]
    if len(data) > 4 and data[:4] == b"%PDF":
        return CONTENT_TYPE_PDF
    if data.startswith(b"<") and b"<html" in data[:1024].lower():
        return CONTENT_TYPE_HTML
    if _looks_like_text(data):
        return CONTENT_TYPE_TEXT
    return "application/octet-stream"


def _looks_like_text(data: bytes) -> bool:
    """Rough text-sniffer that treats high-entropy blobs as binary."""
    if not data:
        return True
    sample = data[:512]
    try:
        sample.decode("utf-8")
    except UnicodeDecodeError:
        return False
    return True


def extract_plain(raw: bytes, content_type: str, extension: str) -> str:
    """Route ``raw`` to the right extractor and return plain text.

    Markdown, text, JSON, YAML round-trip as decoded UTF-8. HTML is
    stripped. PDF is decoded via ``pdfplumber``. Unknown types raise.
    """
    base = _media_type(content_type)
    if base in (
        CONTENT_TYPE_MARKDOWN,
        CONTENT_TYPE_TEXT,
        CONTENT_TYPE_JSON,
        CONTENT_TYPE_YAML,
    ):
        return _decode_utf8(raw)
    if base == CONTENT_TYPE_HTML:
        return strip_html(raw)
    if base == CONTENT_TYPE_PDF:
        return _extract_pdf(raw)
    if extension == ".pdf" or (len(raw) > 4 and raw[:4] == b"%PDF"):
        return _extract_pdf(raw)
    if base.startswith("text/"):
        return _decode_utf8(raw)
    raise ValueError(f"knowledge: unsupported content-type {content_type!r}")


def _decode_utf8(raw: bytes) -> str:
    """Decode ``raw`` as UTF-8 or raise a descriptive error."""
    try:
        return raw.decode("utf-8")
    except UnicodeDecodeError as exc:
        raise ValueError("knowledge: content is not valid UTF-8") from exc


def strip_html(raw: bytes) -> str:
    """Drop script/style blocks, tags, and common entities from ``raw``."""
    try:
        s = raw.decode("utf-8", errors="ignore")
    except Exception:  # pragma: no cover - decode(errors=ignore) never raises
        s = raw.decode("latin-1", errors="ignore")
    s = _SCRIPT_RE.sub(" ", s)
    s = _STYLE_RE.sub(" ", s)
    s = _TAG_RE.sub(" ", s)
    entities = {
        "&nbsp;": " ",
        "&amp;": "&",
        "&lt;": "<",
        "&gt;": ">",
        "&quot;": '"',
        "&#39;": "'",
        "&apos;": "'",
    }
    for entity, replacement in entities.items():
        s = s.replace(entity, replacement)
    return _collapse_whitespace(s)


def _extract_pdf(raw: bytes) -> str:
    """Return the concatenated plain text of every PDF page."""
    if not raw:
        raise ValueError("knowledge: empty pdf body")
    try:
        import pdfplumber
    except ImportError as exc:  # pragma: no cover - optional extra
        raise RuntimeError("knowledge: pdfplumber not installed") from exc

    pieces: list[str] = []
    try:
        with pdfplumber.open(io.BytesIO(raw)) as pdf:
            for page in pdf.pages:
                try:
                    text = page.extract_text() or ""
                except Exception:  # noqa: BLE001 - PDF extractors trip on fonts
                    continue
                text = text.strip()
                if text:
                    pieces.append(text)
    except Exception as exc:  # noqa: BLE001
        raise ValueError(f"knowledge: opening pdf: {exc}") from exc
    return _collapse_whitespace("\n\n".join(pieces))


def build_document(
    req: IngestRequest,
    content_type: str,
    source_label: str,
    extracted: str,
    raw: bytes,
    brain_id: str = "",
) -> Document:
    """Assemble a :class:`Document` from a request plus extracted text."""

    now = datetime.now(timezone.utc)
    title = (req.title or "").strip()
    tags = list(req.tags)

    fm, body = parse_frontmatter(extracted)
    if not title:
        title = fm.title or fm.name or _derive_title(body, source_label)
    summary = fm.summary or fm.description
    if fm.tags:
        tags.extend(fm.tags)

    slug = slugify(title)
    if not slug:
        slug = _hash_slug(raw)

    doc_id = _hash_slug((slug + ":").encode("utf-8") + raw)
    path = raw_document_path(slug)

    return Document(
        id=DocumentID(doc_id),
        brain_id=brain_id or req.brain_id,
        path=path,
        title=title,
        source=source_label,
        content_type=content_type,
        summary=summary,
        body=body.strip(),
        bytes=len(raw),
        tags=_dedupe_strings(tags),
        ingested=now,
        modified=now,
    )


def build_frontmatter_yaml(doc: Document) -> str:
    """Emit the canonical YAML frontmatter block written with each ingest."""
    lines = ["---"]
    lines.append(f'title: {_quote_yaml(doc.title)}')
    if doc.summary:
        lines.append(f'summary: {_quote_yaml(doc.summary)}')
    lines.append(f'source: {_quote_yaml(doc.source)}')
    lines.append(f'source_type: {_quote_yaml(_route_source_type(doc.content_type))}')
    if doc.ingested is not None:
        lines.append(f'ingested: {_quote_yaml(_fmt_rfc3339(doc.ingested))}')
    if doc.modified is not None:
        lines.append(f'modified: {_quote_yaml(_fmt_rfc3339(doc.modified))}')
    if doc.tags:
        lines.append("tags:")
        for tag in doc.tags:
            lines.append(f"  - {tag}")
    lines.append("---")
    return "\n".join(lines)


def normalise_url(raw: str) -> str:
    """Trim, default-to-HTTPS, and validate a URL."""
    raw = raw.strip()
    if not raw:
        raise ValueError("knowledge: empty URL")
    if "://" not in raw:
        raw = "https://" + raw
    parsed = urlparse(raw)
    if not parsed.scheme or not parsed.netloc:
        raise ValueError("knowledge: URL missing host")
    return parsed.geturl()


# --- helpers ---------------------------------------------------------------


def slugify(title: str) -> str:
    """Lowercase, hyphen-separated, 60-char cap slug.

    Matches Go's ``slugify``: Unicode letters and digits survive;
    everything else becomes a hyphen. Runs of hyphens collapse to one.
    """
    s = title.strip().lower()
    pieces: list[str] = []
    for ch in s:
        if _is_letter(ch) or ch.isdigit():
            pieces.append(ch)
        else:
            pieces.append("-")
    slug = "".join(pieces)
    # Collapse consecutive hyphens.
    while "--" in slug:
        slug = slug.replace("--", "-")
    slug = slug.strip("-")
    if len(slug) > 60:
        slug = slug[:60].rstrip("-")
    return slug


def _is_letter(ch: str) -> bool:
    """Report whether ``ch`` is a Unicode letter (matches Go's unicode.IsLetter)."""
    if not ch:
        return False
    return unicodedata.category(ch).startswith("L")


def _hash_slug(data: bytes) -> str:
    """Deterministic 12-char hex fallback slug."""
    return hashlib.sha256(data).hexdigest()[:12]


def _derive_title(body: str, fallback: str) -> str:
    """Pick a best-effort title from ``body`` when the caller gives none."""
    for line in body.split("\n"):
        stripped = line.strip()
        if stripped.startswith("# "):
            return stripped[2:].strip()
    for line in body.split("\n"):
        stripped = line.strip()
        if stripped:
            if len(stripped) > 120:
                stripped = stripped[:120]
            return stripped
    base = _basename(fallback)
    base = _strip_extension(base)
    return base or "untitled"


def _basename(p: str) -> str:
    """POSIX-style basename that also handles Windows separators."""
    for sep in ("/", "\\"):
        idx = p.rfind(sep)
        if idx >= 0:
            p = p[idx + 1 :]
    return p


def _strip_extension(p: str) -> str:
    """Trim the rightmost extension, if any."""
    idx = p.rfind(".")
    if idx > 0:
        return p[:idx]
    return p


def _extension(p: str) -> str:
    """Return the rightmost extension including the dot, lowercased."""
    idx = p.rfind(".")
    if idx < 0:
        return ""
    return p[idx:].lower()


def _collapse_whitespace(s: str) -> str:
    """Trim and collapse whitespace runs while preserving paragraph breaks."""
    paragraphs = s.split("\n\n")
    for i, p in enumerate(paragraphs):
        paragraphs[i] = " ".join(p.split())
    return "\n\n".join(paragraphs).strip()


def _media_type(raw: str) -> str:
    """Normalise ``text/html; charset=utf-8`` to ``text/html``."""
    base = raw.strip().lower()
    idx = base.find(";")
    if idx >= 0:
        base = base[:idx].strip()
    return base


def _dedupe_strings(items: list[str]) -> list[str]:
    """Remove blanks and duplicates while preserving order."""
    seen: set[str] = set()
    out: list[str] = []
    for s in items:
        key = s.strip()
        if not key or key in seen:
            continue
        seen.add(key)
        out.append(key)
    return out


def _route_source_type(content_type: str) -> str:
    """Map a content-type onto the frontmatter source_type tag."""
    mapping = {
        CONTENT_TYPE_MARKDOWN: "markdown",
        CONTENT_TYPE_TEXT: "text",
        CONTENT_TYPE_HTML: "html",
        CONTENT_TYPE_PDF: "pdf",
        CONTENT_TYPE_JSON: "json",
        CONTENT_TYPE_YAML: "yaml",
    }
    return mapping.get(content_type, "document")


def _quote_yaml(value: str) -> str:
    """Double-quote ``value`` and escape embedded quotes/backslashes."""
    escaped = value.replace("\\", "\\\\").replace('"', '\\"')
    return f'"{escaped}"'


def _fmt_rfc3339(when: datetime) -> str:
    """Format a ``datetime`` as RFC 3339 with second precision."""
    if when.tzinfo is None:
        when = when.replace(tzinfo=timezone.utc)
    # Go's time.RFC3339 uses ``2006-01-02T15:04:05Z07:00``. Python's
    # ``isoformat`` matches when we trim microseconds and swap UTC to
    # the ``Z`` shorthand for clarity.
    iso = when.replace(microsecond=0).isoformat()
    if iso.endswith("+00:00"):
        iso = iso[:-6] + "Z"
    return iso


def first_non_empty(*values: str) -> str:
    """Return the first non-blank argument or an empty string."""
    for v in values:
        if v and v.strip():
            return v
    return ""


def read_body(req: IngestRequest) -> tuple[bytes, str, str]:
    """Load the request body and return ``(bytes, content_type, source)``.

    Precedence matches Go: inline ``content`` beats ``path``; when only
    ``path`` is set the file is opened from disk. Returns
    ``content_type`` as the base media type (charset stripped).
    """

    ctype = _media_type(req.content_type)

    if req.content is not None:
        data = _read_reader(req.content)
        if not ctype:
            ctype = detect_content_type(req.path, data)
        return data, ctype, first_non_empty(req.path, "inline")

    path = (req.path or "").strip()
    if not path:
        raise ValueError("knowledge: either content or path required")
    if "://" in path:
        raise ValueError(f"knowledge: use ingest_url for {path}")

    import os

    abs_path = os.path.abspath(path)
    if not os.path.exists(abs_path):
        raise FileNotFoundError(f"knowledge: stat {abs_path}: not found")
    if os.path.isdir(abs_path):
        raise IsADirectoryError(f"knowledge: {abs_path} is a directory")
    size = os.path.getsize(abs_path)
    if size > MAX_READ_BYTES:
        raise ValueError(f"knowledge: {abs_path} exceeds {MAX_READ_BYTES} byte limit")
    with open(abs_path, "rb") as handle:
        data = handle.read()
    if not ctype:
        ctype = detect_content_type(abs_path, data)
    return data, ctype, abs_path


def _read_reader(source: bytes | BinaryIO) -> bytes:
    """Read at most :data:`MAX_READ_BYTES` from ``source``."""
    if isinstance(source, (bytes, bytearray, memoryview)):
        data = bytes(source)
        if len(data) > MAX_READ_BYTES:
            return data[:MAX_READ_BYTES]
        return data
    if hasattr(source, "read"):
        # Accept ``str`` or ``bytes`` readers; encode text defensively.
        raw = source.read(MAX_READ_BYTES + 1)
        if isinstance(raw, str):
            raw = raw.encode("utf-8")
        if raw is None:
            raw = b""
        if len(raw) > MAX_READ_BYTES:
            raw = raw[:MAX_READ_BYTES]
        return raw
    raise TypeError(f"knowledge: unsupported content type {type(source)!r}")


# Expose the Frontmatter type via a re-export for callers that want a
# single import surface from ``knowledge.ingest``. Keeps the Go ergonomics.
__all__.append("Frontmatter")
